templates/hosting/initEmulators.js

```diff
@@ -0,0 +1,31 @@
+if (typeof firebase === 'undefined') {
+  throw new Error('hosting/init-error: Firebase SDK not detected. You must include it before /__/firebase/initEmulators.js');
+}
+
+if (firebase.apps.length === 0) {
+  throw new Error('hosting/init-error: Firebase app not detected. You must call initializeApp before /__/firebase/initEmulators.js');
+}
+
+/*--EMULATORS--*/
+if (firebaseEmulators) {
+  console.warn("Automatically connecting Firebase SDKs to running emulators:");
+  Object.keys(firebaseEmulators).forEach(function(key) {
+    console.warn('\t' + key + ': http://' +  firebaseEmulators[key].host + ':' + firebaseEmulators[key].port );
+  });
+
+  if (firebaseEmulators.database && typeof firebase.database === 'function') {
+    firebase.database().useEmulator(firebaseEmulators.database.host, firebaseEmulators.database.port);
+  }
+
+  if (firebaseEmulators.firestore && typeof firebase.firestore === 'function') {
+    firebase.firestore().useEmulator(firebaseEmulators.firestore.host, firebaseEmulators.firestore.port);
+  }
+
+  if (firebaseEmulators.functions && typeof firebase.functions === 'function') {
+    firebase.functions().useEmulator(firebaseEmulators.functions.host, firebaseEmulators.functions.port);
+  }
+
+  if (firebaseEmulators.auth && typeof firebase.auth === 'function') {
+    firebase.auth().useEmulator('http://' + firebaseEmulators.auth.host + ':' + firebaseEmulators.auth.port);
```

Comment:
Please quote IPv6 hosts like `::1` => `[::1]`

Reply:
Can you explain this one a little more?  Not sure how IPv6 works

Reply:
Basically when constructing URLs, if the hostname contains `:`, the entire hostname must be quoted around `[]` to disambiguate from `:port`. In other words, host `::1` and port `9009` should be encoded as `http://[::1]:9099/`.

This is one of the other reasons why I don't like the API taking URLs, but we gotta do what we have to do.